README.md

# Cadaver: The Rotten Expansion
Cadaver is a mod for The Binding of Isaac: Repentance that adds 11 new unlocks and 1 new playable character (tainted version included)! Explore the basement with Cadaver, Isaac's rotting corpse, and gain access to Golden Souls, Rotten Chests, and more!

## Cadaver
***Base Stats***  
Damage: 6.36  
Fire Rate: 1.5  
Max Health: 0 (+ 1 Bone Heart filled with a Rotten Heart)  
*All other stats are the same as Isaac.*  

***Starting Item***   
Yuck Heart   

***Passives***
* Cadaver only has one rotten eye, so he can only fire at a **fixed 1.5 tears per second**. Those tears are nasty though! All **tear pickups are converted to an equivalent damage multiplier.** Cadaver keeps the same DPS as normal Isaac, but with all of the stats weighted towards Damage.
* Cadaver can **only obtain Rotten Hearts and Bone Hearts.** He can still obtain heart containers, but they must be filled with Rotten Hearts. 
* Items that heal Cadaver or give maximum HP will instead give Rotten Hearts.
* Any **red hearts that spawn are converted** to Rotten Hearts, with a rare chance to convert into a Bone Heart.
* Cadaver **can pickup soul hearts, but they are immediately removed to spawn friendly critters**! You made friends in the grave, after all.
* Blue flies, poison, and some similar effects do *reduced* damage. This is due to a bug in Isaac, but Cadaver is balanced around it.

***Birthright***  
*Better friends*  
When Cadaver has Birthright, all current and future **blue flies are converted into random Locusts.**  

## Tainted Cadaver
***Base Stats***  
Damage: 1.15  
Fire Rate: 5.0  
Max Health: 0 (+ 2 Bone Hearts filled with a Rotten Heart)  
*All other stats are the same as Isaac.*  

***Starting Item***   
Halitosis (see details below)   

***Stat Passives***
* Tainted Cadaver fires at a **fixed 5.0 tears per second.** All **tear pickups are converted to an equivalent damage multiplier.**
* Tainted Cadaver has **40% reduced damage.**
* Tainted Cadaver can **only obtain Bone Hearts and fill them with Rotten Hearts.**
* Tainted Cadaver can only have a **maximum of four bone hearts.**
* Items that heal Tainted Cadaver or give maximum HP will instead give Bone + Rotten Hearts.
* Any **red hearts that spawn are converted** to Rotten Hearts, with a small chance to convert into a Bone Heart.
* Poison, blue spiders, and some similar effects do *increased* damage. This is due to a bug in Isaac, but Tainted Cadaver is balanced around it.

***Friendly Critters***
* Tainted Cadaver **can pickup soul hearts, but they are immediately removed to spawn friendly critters**!
* Tainted Cadaver's friendly **critters are permanent and deal 3.0 damage**
* Using **Halitosis will respawn any dead critters** and will relocate all alive critters to Tainted Cadaver's location.
* If Tainted Cadaver has BFFS!, critters will do 4.5 damage.

***Birthright***  
*Respawning friends*  
When Tainted Cadaver has Birthright, his **friendly units automatically respawn after 3 seconds.**  

## Unlocks
### Vestments [Angel Room]
*Defeat Isaac and ??? as Cadaver*  
Gain a random Angel Room item effect in every room. This item counts towards the Seraphim transformation. Leaving and re-entering a room will give the same effect each time.

### Probiotics [Trinket]
*Defeat Satan and The Lamb as Cadaver*  
If Isaac has a Rotten Heart, gain a moderate stat boost:  
+1 flat damage  
-0.5 Tear Delay  
+0.25 Speed  
+1 Luck

### Soul of Cadaver
*Defeat Hush as Cadaver*  
Spawns 8 small friendly leeches, 8 small friendly maggots, 1 bone heart, and 2 rotten hearts.

### Cadaver's Left Eye [Treasure Room]
*Defeat Delirium as Cadaver*  
+2 flat damage  
Fixes fire rate at 1.5 and converts all tear ups into an equivalent damage multiplier.

### Forbidden Fruit [Devil Room]
*Defeat Mother as Cadaver*  
One time use active. When used, gain a permanent 3x damage multiplier and 0.5x tear delay multiplier. Once used, Isaac takes 1 instance of damage every 4 seconds until the end of the run (persists through resurrections).

### Golden Soul
*Defeat The Beast as Cadaver*  
Activates a random soul effect, excluding Soul of Lazarus. Has a 60% chance to spawn another Golden Soul.

### Rotten Chests
*Defeat Mega Satan as Cadaver*  
Rotten Chests replace 5% of normal chest spawns. They do not require a key.  
When opened, has a 15% chance to spawn a "rotten" pedestal item. Otherwise, spawns 2 normal pickups and 1 "rotten" pickup.  

Normal pickups: coin, heart, key, bomb.  
Rotten pickups: 2 rotten hearts, 1 bone heart, 1 "rotten" trinket  

### Halitosis [Treasure Room]
*Defeat Hush and Delirium as Tainted Cadaver*  
Active; 8 second cooldown. Fire 6 large breath tears over 1 second. These tears pierce, apply poison and fear, and deal Isaac's tear damage.  
When used as Tainted Cadaver, respawns all allied critters at Tainted Cadaver's location.  
  
If Isaac has Car Battery, Halitosis fires 12 tears over 2 seconds.  

### Morgue Key [Secret Room]
*Defeat Isaac, ???, Satan, The Lamb, and Mega Satan as Tainted Cadaver*  
One time use active. When used, teleport to the Morgue.  
The Morgue includes a collection of pedestal items from the last completed run. Only one can be taken.  

### Tech Drones [Treasure Room]
*Defeat Mother as Tainted Cadaver*  
Gain two familiar robots that orbit Isaac at a large distance.  
When Isaac fires tears, a laser connects the two bots, dealing 40% of Isaac's tear damage per tick.  
  
If Isaac has BFFS!, the laser does 60% of Isaac's tear damage.  
If Isaac uses Box of Friends, the bots are duplicated temporarily.  

### Hydrochloric Acid [Treasure Room]
*Defeat Beast as Tainted Cadaver*  
After hitting an enemy 5 times, all future attacks will deal an additional 3.0 flat damage.  
Synergizes well with Brimstone, Technology, and other weapons that have a high tick rate.  

## Multiplayer Interactions
* Item unlock progression is only tracked if Player 1 is Cadaver/Tainted Cadaver.
* Morgue Key will only remember items held by Player 1
* Heart conversions to Rotten + Bone hearts are only done if Player 1 is Cadaver/Tainted Cadaver.
* Multiple Tainted Cadavers share one army that respawns when any Tainted Cadaver uses Halitosis.
* All other interactions should now work as would be expected, but limited testing has been done.

## Known Issues
* Completion marks do not display on the main menu, even though they are being tracked. There are other mods to add completion marks to modded characters.
* Several effects ignore item unlocks. For instance, you can obtain bone hearts before unlocking them.
* Achievements show up immediately upon a boss dying, rather than after the run has completed / in between stages.


changelog.txt
2.9
===
- Fixed crash when continuing a Tainted Cadaver run after restarting the game

2.5
===
- Fixed Morgue Key not working in Greed mode

2.4
===
- New sprites for items, objects, and more!
- Fix soul sprites when held as a secondary pocket item
- Fix Vestments removing transformation progress
- Fix Guppy transformation visuals

2.3
===
- Several co-op fixes
- Forbidden Fruit has been added to the demon beggar pool
- Halitosis and Cadaver's Left Eye have been added to the rotten beggar pool
- All items have been added to the appropriate greed pools
- Removed the Soul of Lost from the Golden Soul pool
- Rune Shards will no longer spawn when Soul of Cadaver and Golden Souls have not been unlocked
- Halitosis now has custom sfx
- Halitosis now overrides Trisagion during the active
- The Morgue can now contain Bag of Crafting, Sumptorium, etc.
- Rotten Chest pedestal items now use the Rotten Beggar pool
- Rotten Chests can no longer spawn in challenge rooms
- When Cadaver has Birthright and Rotten Baby, the player is limited to 10 total locusts at a time (it's still really strong)
- Vestments now properly gives the same item in a room no matter where you entered from

2.2
===
Box of Friends no longer always gives Tech Drones (oops)

2.1
===
+ Added EID Item Descriptions
- Tweaked item qualities based on more playtesting
- Fixed scenarios Probiotics effect would not update properly
- Fixed a buggy interaction between Cadaver's Left Eye and Tainted Cadaver's passive. Now it just gives a damage up.
- Hive Mind now properly functions on Tainted Cadaver
- Eternal hearts no longer heal an additional heart when completed on Tainted Cadaver.

2.0
===
Tainted Cadaver Update! See the mod description for the full details.


main.lua
local Achievements = include("Achievements")
local RottenFlesh = include("RottenFlesh")
local TechDrones = include("TechDrones")
local HydrochloricAcid = include("HydrochloricAcid")
local Vestments = include("Vestments")
local ForbiddenFruit = include("ForbiddenFruit")
local Probiotics = include("Probiotics")
local RottenIsaac = include("RottenIsaac")
local Halitosis = include("Halitosis")
local TaintedCadaver = include("TaintedCadaver")
local MorgueKey = include("MorgueKey")
local CustomSouls = include("CustomSouls")  
local RottenChest = include("RottenChest")
local ItemPools = include("ItemPools")
local json = include("json")
local Cadaver = RegisterMod("Cadaver", 1)

Achievements.RegisterCallbacks(Cadaver)

CadaverRNG = RNG() -- Used for general RNG, just seeded at the start with the run seed
CadaverItemRNG = RNG() -- Used for seeded RNG, never use without setting the seed right beforehand

-- # POST GAME START #
function Cadaver:StartRun(isContinued)
  CadaverRNG:SetSeed(Game():GetSeeds():GetStartSeed(), 0)
  
  if Cadaver:HasData() then
    local table = json.decode(Cadaver:LoadData())
    ForbiddenFruit.LoadData(table.ForbiddenFruit)
    Vestments.LoadData(table.Vestments)
    Achievements.LoadData(table.CadaverAchievements)
    TaintedCadaver.LoadData(table.TaintedCadaver)
    MorgueKey.LoadData(table.MorgueKey)
  else
    Achievements.Reset()
  end
  
  Vestments.Reset(isContinued)
  ForbiddenFruit.Reset(isContinued)
  TaintedCadaver.Reset(isContinued)
  MorgueKey.Reset(isContinued)
  Probiotics.Reset(isContinued)

  Achievements.SetupItemPools()
  
  -- Evaluate stats after everything has been setup
  local player = Isaac.GetPlayer(0)
  player:AddCacheFlags(CacheFlag.CACHE_ALL)
  player:EvaluateItems()
end
Cadaver:AddCallback(ModCallbacks.MC_POST_GAME_STARTED, Cadaver.StartRun)

-- # POST NORMAL CHEST CREATION #
function Cadaver:ModifyPickups(pickup)
  RottenChest.ReplaceChests(pickup)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PICKUP_INIT, Cadaver.ModifyPickups, PickupVariant.PICKUP_CHEST)

-- # POST TAROT CARD CREATION #
function Cadaver:ModifyTarotCards(pickup)
  CustomSouls.RemoveLockedSouls(pickup)
  CustomSouls.ChangeSprites(pickup)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PICKUP_INIT, Cadaver.ModifyTarotCards, PickupVariant.PICKUP_TAROTCARD)

-- # POST HEART CREATION #
function Cadaver:ModifyHearts(pickup)
  RottenIsaac.ReplaceHearts(pickup)
  TaintedCadaver.ReplaceHearts(pickup)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PICKUP_INIT, Cadaver.ModifyHearts, PickupVariant.PICKUP_HEART)

-- # EFFECT UPDATES (Every frame) #
function Cadaver:EffectUpdate(player)
  Vestments.UpdateVestmentsEffect(player)
  ForbiddenFruit.UpdateForbiddenFruitEffect(player)
  Probiotics.UpdateProbioticsEffect(player)
  Halitosis.UpdateHalitosisEffect(player)
  MorgueKey.UpdateMorgueKeyEffect(player)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PEFFECT_UPDATE, Cadaver.EffectUpdate)

-- # CADAVER-SPECIFIC EFFECT UPDATES #
function Cadaver:CadaverEffectUpdate(player)
  RottenIsaac.ConvertHealth(player)
  RottenIsaac.Birthright(player)
  RottenIsaac.ManageCostumes(player)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PEFFECT_UPDATE, Cadaver.CadaverEffectUpdate, PlayerType.PLAYER_CADAVER)

-- # TAINTED CADAVER EFFECT UPDATES #
function Cadaver:TaintedCadaverEffectUpdate(player)
  TaintedCadaver.ConvertHealth(player)
  TaintedCadaver.ManageArmy(player)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PEFFECT_UPDATE, Cadaver.TaintedCadaverEffectUpdate, PlayerType.PLAYER_TAINTED_CADAVER)

-- # TECH DRONES UPDATE #
function Cadaver:TechDroneUpdate(familiar)
  TechDrones.Update(familiar)
end
Cadaver:AddCallback(ModCallbacks.MC_FAMILIAR_UPDATE, Cadaver.TechDroneUpdate, FamiliarVariant.TECH_DRONES)

-- # ENTITY TAKES DAMAGE #
function Cadaver:TaintedCadaverArmyDamage(entity, amount, flags, source, countdownFrames)
  if not TaintedCadaver.SoldierDamage(entity, amount, flags, source, countdownFrames) then return false end
  if not TechDrones.LaserDamage(entity, amount, flags, source, countdownFrames) then return false end
  
  HydrochloricAcid.ItemDamage(entity, amount, flags, source, countdownFrames)
  return nil
end
Cadaver:AddCallback(ModCallbacks.MC_ENTITY_TAKE_DMG, Cadaver.TaintedCadaverArmyDamage)

-- # POST LASER INIT #
function Cadaver:PostLaserInit(laser)
  TaintedCadaver:InitSoldierLasers(laser)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_LASER_INIT, Cadaver.PostLaserInit)

-- # POST PLAYER INIT #
function Cadaver:PlayerInit(player)
  if player:GetPlayerType() == PlayerType.PLAYER_CADAVER then
    RottenIsaac.AddCostume(player)
  end
  
  if player:GetPlayerType() == PlayerType.PLAYER_TAINTED_CADAVER then
    TaintedCadaver.PlayerInit(player)
  end
end
Cadaver:AddCallback(ModCallbacks.MC_POST_PLAYER_INIT, Cadaver.PlayerInit)

-- # STAT UPDATES #
function Cadaver:ModifyStats(player, cacheFlag)
  Probiotics.ModifyStats(player, cacheFlag)
  ForbiddenFruit.ModifyStats(player, cacheFlag)
  RottenFlesh.ModifyStats(player, cacheFlag)
  RottenIsaac.ModifyStats(player, cacheFlag)
  TaintedCadaver.ModifyStats(player, cacheFlag)
  TechDrones.UpdateCache(player, cacheFlag)
  Halitosis.ModifyStats(player, cacheFlag)
end
Cadaver:AddCallback(ModCallbacks.MC_EVALUATE_CACHE, Cadaver.ModifyStats)

-- # ROOM ENTER #
function Cadaver:EnterRoom()
  Vestments.SwapVestmentsItem()
  RottenChest.RemoveOpenChests()
  MorgueKey.InitRoom()
end
Cadaver:AddCallback(ModCallbacks.MC_POST_NEW_ROOM, Cadaver.EnterRoom)

-- # RENDER NPC #
function Cadaver:NpcRender(npc, renderOffset)
  HydrochloricAcid.Render(npc, renderOffset)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_NPC_RENDER, Cadaver.NpcRender)

-- # UPDATE NPC #
function Cadaver:NpcUpdate(npc)
  HydrochloricAcid.Update(npc)
end
Cadaver:AddCallback(ModCallbacks.MC_NPC_UPDATE, Cadaver.NpcUpdate)

-- # RENDER #
function Cadaver:Render()
  Achievements.DisplaySprites()
end
Cadaver:AddCallback(ModCallbacks.MC_POST_RENDER, Cadaver.Render)

-- # UPDATE #
function Cadaver:Update()
  Achievements.UpdateSprites()
end
Cadaver:AddCallback(ModCallbacks.MC_POST_UPDATE, Cadaver.Update)

-- # GAME EXIT #
function Cadaver:Exit(shouldSave)
  Vestments.Exit()
  TaintedCadaver.Exit()
  local saveData = {
    ForbiddenFruit = ForbiddenFruit.SaveData(),
    Vestments = Vestments.SaveData(),
    CadaverAchievements = Achievements.SaveData(),
    TaintedCadaver = TaintedCadaver.SaveData(),
    MorgueKey = MorgueKey.SaveData()
  }
  Cadaver:SaveData(json.encode(saveData))
end
Cadaver:AddCallback(ModCallbacks.MC_PRE_GAME_EXIT, Cadaver.Exit)

-- # POST GAME END #
function Cadaver:PostGameEnd(isGameOver)
  MorgueKey.SaveItems(isGameOver)
end
Cadaver:AddCallback(ModCallbacks.MC_POST_GAME_END, Cadaver.PostGameEnd)

-- ###### SPECIFIC ITEMS ######

-- # GOLDEN SOUL USAGE #
function Cadaver:UseGoldenSoul(card, player, flags)
  CustomSouls.UseGoldenSoul(player)
end
Cadaver:AddCallback(ModCallbacks.MC_USE_CARD, Cadaver.UseGoldenSoul, Card.CARD_SOUL_GOLDEN)

-- # CADAVER SOUL USAGE #
function Cadaver:UseCadaverSoul(card, player, flags)
  CustomSouls.UseCadaverSoul(player)
end
Cadaver:AddCallback(ModCallbacks.MC_USE_CARD, Cadaver.UseCadaverSoul, Card.CARD_SOUL_CADAVER)

-- # FORBIDDEN FRUIT USAGE #
function Cadaver:UseForbiddenFruit()
  return ForbiddenFruit.Use()
end
Cadaver:AddCallback(ModCallbacks.MC_USE_ITEM, Cadaver.UseForbiddenFruit, CollectibleType.COLLECTIBLE_FORBIDDEN_FRUIT)

-- # HALITOSIS USAGE #
function Cadaver:UseHalitosis(collectibleType, RNG, player, useFlags, slot)
  Halitosis.Use(collectibleType, RNG, player, useFlags, slot)
  TaintedCadaver.KillArmy()
end
Cadaver:AddCallback(ModCallbacks.MC_USE_ITEM, Cadaver.UseHalitosis, CollectibleType.COLLECTIBLE_HALITOSIS)

-- # MORGUE KEY USAGE #
function Cadaver:UseMorgueKey(collectibleType, RNG, player, useFlags, slot)
  return MorgueKey.Use(collectibleType, RNG, player, useFlags, slot)
end
Cadaver:AddCallback(ModCallbacks.MC_USE_ITEM, Cadaver.UseMorgueKey, CollectibleType.COLLECTIBLE_MORGUE_KEY)

-- # ROTTEN CHEST COLLISION #
function Cadaver:RottenChestCollision(pickup, collider, low)
  RottenChest.OpenChest(pickup, collider, low)
end
Cadaver:AddCallback(ModCallbacks.MC_PRE_PICKUP_COLLISION, Cadaver.RottenChestCollision, PickupVariant.PICKUP_ROTTENCHEST)

-- # DEBUG COMMANDS #
function Cadaver:OnCommand(command, args)
  if command == "cd_reset" then
    Achievements.Reset()
  elseif command == "cd_unlockall" then
    Achievements.UnlockAll()
  elseif command == "cd_listunlock" then
    Achievements.ReadOut()
  elseif command == "cd_army" then
    TaintedCadaver.GetArmyStats()
  elseif command == "cd_entities" then
    local entities = Isaac:GetRoomEntities()
    for i, entity in ipairs(entities) do
      Isaac.DebugString("Type " .. entity.Type .. " Variant " .. entity.Variant .. " SubType " .. entity.SubType)
    end
  end
end
Cadaver:AddCallback(ModCallbacks.MC_EXECUTE_CMD, Cadaver.OnCommand)

-- EID Descriptions
if EID then
  EID:addCollectible(CollectibleType.COLLECTIBLE_FORBIDDEN_FRUIT, "!!! SINGLE USE#↑ 3x Damage up#↑ 2x Tears up#You take damage every 4 seconds#!!! This CANNOT be cured, even by resurrection, and WILL kill you!")
  EID:addCollectible(CollectibleType.COLLECTIBLE_ROTTEN_FLESH, "↑ +2 Damage up#Fixes tear rate at 1.5#All tear rate modifiers are converted into an equivalent damage multiplier")
  EID:addCollectible(CollectibleType.COLLECTIBLE_VESTMENTS, "Grants a random Angel Room item effect each room")
  EID:addCollectible(CollectibleType.COLLECTIBLE_HALITOSIS, "Fires a barrage of large breath tears that pierce and inflict Poison and Fear")
  EID:addCollectible(CollectibleType.COLLECTIBLE_MORGUE_KEY, "!!! SINGLE USE#Teleports you to the Morgue, a special room#It contains 4-5 items from the last successful run on pedestals#Only one can be taken")
  EID:addCollectible(CollectibleType.COLLECTIBLE_TECH_DRONES, "Grants two robot familiars that orbit far away from you, opposite each other#When firing tears, a Technology laser connects the two")
  EID:addCollectible(CollectibleType.COLLECTIBLE_HYDROCHLORIC_ACID, "Hitting an enemy 5 times causes them to take an additional +3 flat damage per hit")
  EID:addTrinket(TrinketType.TRINKET_PROBIOTICS, "If you have a Rotten Heart:#↑ +1 Damage up#↑ +0.5 Tears up#↑ +0.25 Speed up#↑ +1 Luck up")
  EID:addCard(Card.CARD_SOUL_CADAVER, "Spawns:#8 friendly Small Maggots#8 friendly Small Leeches#1 Bone Heart#2 Rotten Hearts")
  EID:addCard(Card.CARD_SOUL_GOLDEN, "Random soul stone effect (except Soul of Lazarus)#60% chance to spawn itself again")
  EID:addBirthright(PlayerType.PLAYER_CADAVER, "Blue flies are converted to random locusts", "Cadaver")
  EID:addBirthright(PlayerType.PLAYER_TAINTED_CADAVER, "Dead familiars respawn after 3 seconds.", "Tainted Cadaver")
end
